refactor(sourcing): move sourcing functionality to `pop-common`

peterwht opened the pull request:
Sourcing functionality is used by both `pop-parachains` and, now, `pop-contracts` for `substrate-contracts-node` .So, this PR refactors the `Source` and `Binary` structs (and related components) into `pop-common` for better usability and reduced duplication.

Review https://github.com/r0gue-io/pop-cli/pull/254 first. But merge this PR first.

evilrobot-01 commented on crates/pop-common/src/sourcing/binary.rs:
```diff
@@ -0,0 +1,253 @@
+use crate::sourcing::{
```
Missing license. Please double check all new files added.

evilrobot-01 commented on crates/pop-common/src/sourcing/binary.rs:
```diff
@@ -0,0 +1,253 @@
+
+/// A descriptor of a remote repository.
+#[derive(Debug, PartialEq)]
+pub struct Repository {
```
Debatable on whether this should have been moved as its currently only being used by the parachains crate. Probably fine. Questionable whether it should be in the binary module, apart from the package field which links to a buildable Rust package which _may_ be a binary.

evilrobot-01 commented on crates/pop-common/src/sourcing/binary.rs:
```diff
@@ -0,0 +1,253 @@
+}
+
+/// Trait for observing status updates.
+pub trait Status {
```
This should be in the crate root imo, its a general purpose trait for reporting status updates and not specific to a binary. Likewise the default impl for ().

evilrobot-01 commented on crates/pop-common/src/sourcing/binary.rs:
```diff
@@ -0,0 +1,253 @@
+}
+
+/// Determines the target triple based on the current platform.
+pub fn target() -> Result<&'static str, Error> {
```
Should be in sourcing/mod.rs, or higher.

evilrobot-01 commented on crates/pop-common/src/sourcing/mod.rs:
```diff
@@ -0,0 +1,3 @@
+pub mod binary;
+pub mod errors;
+pub mod sourcing;
```
Think this adds unneccessary complexity/duplication - sourcing::sourcing::*. Suggest moving the content of sourcing.rs to sourcing/mod.rs so its easier to find and easier to navigate.

evilrobot-01 commented on crates/pop-contracts/Cargo.toml:
```diff
@@ -36,7 +36,6 @@ contract-extrinsics.workspace = true
 
 #  pop
 pop-common = { path = "../pop-common", version = "0.2.0" }
-pop-parachains = { path = "../pop-parachains", version = "0.2.0" }
```
wtf

peterwht replied:
This was an intermediate implementation that I used in PR 254.

evilrobot-01 commented on crates/pop-cli/tests/contract.rs:
```diff
@@ -43,8 +44,11 @@ async fn contract_lifecycle() -> Result<()> {
-	let cache = temp_dir.join("cache");
-	let process = run_contracts_node(cache, None).await?;
+	// temp_dir gets dropped prematurely, so manually create a tmp directory.
+	let cache = dirs::cache_dir().expect("cache_dir failed").join("pop_tmp");
```
This seems weird as it should only drop when `temp` goes out of scope.

peterwht replied:
It's because the temp_dir handle gets moved in the subsequent functioned calls, so, although the original handle still exists (when using clone), the clones get dropped which result in the directory getting prematurely cleaned.

Passing references introduces unnecessary complexity in terms of manually specifying lifetimes on the `Binary` struct. So, I felt manually creating a temp_dir is the least overhead for a test.

evilrobot-01 commented on crates/pop-cli/src/commands/up/parachain.rs:
```diff
@@ -8,7 +8,7 @@ use cliclack::{
 use console::{Emoji, Style, Term};
 use duct::cmd;
-use pop_common::sourcing::binary::Status;
+use pop_common::sourcing::Status;
```
As suggested elsewhere, Status could/should be moved to root module as its fairly abstract to general purpose status reporting.

evilrobot-01 commented on crates/pop-common/src/sourcing/errors.rs:
```diff
@@ -0,0 +1,23 @@
+// SPDX-License-Identifier: GPL-3.0
```
A whole module for a single enum seems unnecesssary, especially when it is named errors but there is/will probably only ever be a single error type. This pattern has developed from previous implementations and not sure it should be followed. May be subjective however.

evilrobot-01 reviewed (changes requested):
Looks great, thanks for taking this on! Have made some suggestions on improving the placement of things, mostly for visibility/accessibility for crate users. Should be quick to address if no objections.

peterwht commented:
- Moved `Status` to `pop_common -> lib.rs`
- Moved `Repository` to `pop_common -> git`
- Moved sourcing error to `sourcing -> mod.rs`
- Moved `target()` to `common -> helpers`

AlexD10S reviewed (approved):
LGTM!
